### PR TITLE
Added tests and modules for mixin, rule stmt and quantifier expressions

### PR DESCRIPTION
**Pull request by Vishalk91-4**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y

 re #8 

#### 2. What is the scope of this PR (e.g. component or file name):

- grammar.js
- expr.txt
- schema.txt
- grammar.json
- node-types.json
- parser.c
- parser.h

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [x] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

This PR does the below tasks

- [x] Added the mixin_statement
- [x] Added Test Cases for rule_statement
- [x] Added modules of quantifier expression in support of PR #11 and added it to parser
- [x] Added modules for optional attributes in support of PR #13 and added it to parser
- [x] Refined and added more test cases for quantifier expression
- [x] Added Protocol and Check in Schema with their module defination

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [ ] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by Peefy** [comment]
PR confilct

**Comment by Vishalk91-4** [comment]
> PR confilct

Resolved PR conflict occuring

File: test/corpus/expr.txt [hunk 1]
```diff
@@ -165,6 +165,46 @@ b = data?.name
...
+
+--------------------------------------------------------------------------------
+
+(module
+      (assignment
+        (dotted_name
+          (identifier))
+        (comparison_operator
+          (identifier)
+          (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: Unexpected `ERRPR` token.

File: test/corpus/expr.txt [hunk 2]
```diff
@@ -165,6 +165,46 @@ b = data?.name
...
+                (string_end)))
+            (pair
+              (identifier)
+              (string
+                (string_start)
+                (string_content)
+                (string_end))))))
+      (MISSING _newline)
+      (dictionary
+        (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: Unexpected ERROR token.

**Comment by Peefy** [comment]
Add some comments and CI failed.

File: test/corpus/expr.txt [hunk 3]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+================================================================================
+quantifier all expression
+================================================================================
+
+all x in [1, 2, 3, 4] { x > 0 }
+
+--------------------------------------------------------------------------------
+
+(module
+  (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 4]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+    (quant_op))
+  (comparison_operator
+    (identifier)
+    (list
+      (integer)
+      (integer)
+      (integer)
+      (integer)))
+  (MISSING _newline)
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 5]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+  (dictionary
+    (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 6]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+================================================================================
+quantifier any expression
+================================================================================
+
+any x in [1, 2, 3, 4] { x == 2 }
+
+--------------------------------------------------------------------------------
+
+(module
+  (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 7]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+            (string_start)
+            (string_content)
+            (string_end)))
+        (pair
+          (identifier)
+          (string
+            (string_start)
+            (string_content)
+            (string_end))))))
+  (MISSING _newline)
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 8]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+
+d = filter _, v in {a = "foo", b = "bar"} {v == "foo"}
+
+--------------------------------------------------------------------------------
+
+(module
+  (assignment
+    (dotted_name
+      (identifier))
+    (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 9]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+            (string_start)
+            (string_content)
+            (string_end)))
+        (pair
+          (identifier)
+          (string
+            (string_start)
+            (string_content)
+              (string_end))))))
+        (MISSING _newline)
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

File: test/corpus/expr.txt [hunk 10]
```diff
@@ -739,3 +701,128 @@ Braces expressions
...
+        (dictionary
+        (ERROR
```
Comment (Peefy) [inline test/corpus/expr.txt]: unexpected `(ERROR` token.

**Comment by Vishalk91-4** [comment]
Will fix all unwanted ERROR codes 
And push a commit

**Comment by Vishalk91-4** [comment]
@Peefy 
I found out that the previous `PRs merged` did not run `tree-sitter generate` while adding things to grammar.js and so none  of them were parsed, they were just added as `ERROR` token or something else.
And directly ran the `tree-sitter test` which also had a lot of `ERROR` tokens in it and passed it

I'll fix them all and create a commit soon here

**Comment by Vishalk91-4** [comment]
@Peefy 

The previous PR #14 merged, added `number_bin_suffix_expr`, and so now for most of test cases, I need to change `integer` to this `number_bin_suffix_expr` and `list` to `braces_exprerssion`
Don't you think it would cause difficulty in understanding as `integer`, `list` previously was much readable when seen in tests

**Comment by Peefy** [comment]
> @Peefy
> 
> The previous PR #14 merged, added `number_bin_suffix_expr`, and so now for most of test cases, I need to change `integer` to this `number_bin_suffix_expr` and `list` to `braces_exprerssion` Don't you think it would cause difficulty in understanding as `integer`, `list` previously was much readable when seen in tests

I think `integer` and `list` are better.

**Comment by Vishalk91-4** [comment]
> > @Peefy
> > The previous PR #14 merged, added `number_bin_suffix_expr`, and so now for most of test cases, I need to change `integer` to this `number_bin_suffix_expr` and `list` to `braces_exprerssion` Don't you think it would cause difficulty in understanding as `integer`, `list` previously was much readable when seen in tests
> 
> I think `integer` and `list` are better.

So do you want me to remove the `number_bin_suffix_expr`, as it's presence will force me to add it in place of `integer` for test cases

**Comment by Peefy** [comment]
> > > @Peefy
> > > The previous PR #14 merged, added `number_bin_suffix_expr`, and so now for most of test cases, I need to change `integer` to this `number_bin_suffix_expr` and `list` to `braces_exprerssion` Don't you think it would cause difficulty in understanding as `integer`, `list` previously was much readable when seen in tests
> > 
> > 
> > I think `integer` and `list` are better.
> 
> So do you want me to remove the `number_bin_suffix_expr`, as it's presence will force me to add it in place of `integer` for test cases

Yes.

**Comment by Vishalk91-4** [comment]
@Peefy 
I was just asking, did we need that most recent `PR Merged` in `dotted_name: $ => prec(1, sep1($.identifier, choice('?.','.',))),` when we already have `attribute` and `optonal_attribute` and `optional_item`

**Comment by Peefy** [comment]
No

**Comment by Vishalk91-4** [comment]
> No

So do you want me to keep that change of ` dotted_name: $ => prec(1, sep1($.identifier, choice('?.','.',))),` or remove them and we can use the `optional_attribute` and `optional_item`

**Comment by Vishalk91-4** [comment]
@Peefy 
Could you review it now, I have removed the `Error` codes in most I guess

**Review by Peefy — APPROVED** [review]
LGTM

**Pull request merged**